### ISSUE TITLE
qsoqn, rrdesi, multiprocessing, and perlmutter incompatibilities

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#1902]
I suspect that @akremin has tripped on this while developing the desi_zproc branch:

The `desispec.scripts.qsoqn` afterburner calls `redrock.external.desi.rrdesi`, but that spins off 64 multiprocessing ranks to run redrock.  Two problems with this:
* in desi_zproc, the call to qsoqn -> rrdesi is done by an MPI rank, and MPI gets jealous about anyone else trying to do parallelism and can hang.  We've been lucky about this on Cori, but apparently it just won't work on Perlmutter (NERSC staff tried hard, including iterating with Cray/HPE, and in the end the conclusion was that Perlmutter just won't support MPI+multiprocessing).
* even if it did support MPI+multiprocessing, we have 10 ranks spawning 64 multiprocessing processes each (log message "Running with 64 processes"), way over-subscribing the physical cores.

Options:
* update rrdesi so that --mp 1 turns off multiprocessing (unfortunately, unlike most of our multiprocessing enabled scripts, setting the number of multiprocessing processes to 1 in rrdesi continues to use multiprocessing, albeit just spawning 1 process which is still problematic for MPI)
* update desi_zproc to pass an MPI sub-communicator into the qsoqn afterburner, and have it call rrdesi(..., comm) instead.  This is likely easier than fully disabling both MPI and multiprocessing within redrock, but requires the usual mess of "if rank == 0" stuff so that you aren't running the rest of the afterburner N>>1 times.
* like the previous, but use a subcommunicator of size=1.  That wouldn't require any other updates to the qsoqn afterburner itself, and if running redrock in GPU mode it may not matter that it only gets a single CPU rank (though in this case, there is also the risk that all 10 subcommunicators will try to use GPU=0, leaving the others idle).

Note: this is likely a fragility of the current coadd-redshifts bash script too, which still runs `srun -N 1 -n 1 -c 128 --cpu-bind=none desi_qso_qn_afterburner ...` to get each petal running on a different node, but then that afterburner is calling redrock in multiprocessing mode from an MPI rank.

@akremin @marcelo-alvarez @dmargala @craigwarner-ufastro @lastephey

**Comment by moustakas** [comment]
> * update rrdesi so that --mp 1 turns off multiprocessing (unfortunately, unlike most of our multiprocessing enabled scripts, setting the number of multiprocessing processes to 1 in rrdesi continues to use multiprocessing, albeit just spawning 1 process which is still problematic for MPI)

I'm a distant observer on this ticket but I wanted to quickly mention that I've run into the issue above many times and so my adopted solution is to always capture the `mp=1` case. Here's a code snippet from FastSpecFit, for example (noting that `Spec.ntargets` is always a reasonable number, no more than 1000 or so):
```
    fitargs = [(iobj, data[iobj], out[iobj], meta[iobj], CFit, EMFit, args.verbose, args.broadlinefit)
               for iobj in np.arange(Spec.ntargets)]
    if args.mp > 1:
        import multiprocessing
        with multiprocessing.Pool(args.mp) as P:
            _out = P.map(_fastspec_one, fitargs)
    else:
        _out = [fastspec_one(*_fitargs) for _fitargs in fitargs]
```

**Comment by sbailey** [comment]
Update: redrock uses a more complex multiprocessing method of spawning processes that use Queue objects to pass results back, so it isn't trivial to scrub multiprocessing out of the n=1 case.

However, it wasn't too hard to add a size=1 communicator to the zproc -> qsoqn -> redrock path in PR #1904, so this issue will be fixed when that PR is merged (and it already works in that branch).

**Comment by sbailey** [comment]
Fix in PR #1904 with the new `desi_zproc` script.  Closing.
